### PR TITLE
Test build hooks

### PR DESCRIPTION
**Pull request by mosuem**
A first draft for how a build hook test might look like.

cc @dcharkes 

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: pkgs/native_assets_cli/example/build/local_asset/test/build_test.dart [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:native_assets_cli/native_assets_cli_test.dart';
+import 'package:test/test.dart';
+
+import '../hook/build.dart' as build;
+
+void main() {
+  testBuild(
```
Comment (dcharkes) [inline pkgs/native_assets_cli/example/build/local_asset/test/build_test.dart:11]: `testBuildHook`? That seems more clear.

**Review by dcharkes — COMMENTED** [review]
This is a great idea! I left a bunch of comments on which direction we could go in: providing more specific `expect`s, having a list of default checks (that the files exist for example).

This also neatly ties in with the `validate` function we should write, we can add the same checks by default to the `testBuildHook`!

File: pkgs/native_assets_cli/lib/native_assets_cli_test.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export 'src/api/build_output.dart' show BuildOutput, LinkOutput;
+export 'src/api/builder.dart' show Builder;
+export 'src/api/hook_config.dart' show HookConfig;
+export 'src/api/ios_sdk.dart' show IOSSdk;
+export 'src/api/link.dart' show link;
+export 'src/api/link_config.dart' show LinkConfig;
+export 'src/api/link_mode_preference.dart' show LinkModePreference;
+export 'src/api/linker.dart' show Linker;
+export 'src/api/os.dart' show OS;
+export 'src/api/test.dart' show testBuild;
```
[the remark this answers is not in the document]
Comment (dcharkes) [reply pkgs/native_assets_cli/lib/native_assets_cli_test.dart:33]: Maybe also check the `dart doc` results of re-exporting

File: pkgs/native_assets_cli/lib/native_assets_cli_test.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
```
[the remark this answers is not in the document]
Comment (mosuem) [reply pkgs/native_assets_cli/lib/native_assets_cli_test.dart:1]: As always, this surfaces my copy-pasting attempts ;)

File: pkgs/native_assets_cli/pubspec.yaml [hunk 4]
```diff
@@ -21,11 +21,12 @@ dependencies:
   collection: ^1.17.1
   crypto: ^3.0.3
   logging: ^1.2.0
+  meta: ^1.15.0
   pub_semver: ^2.1.3
-  yaml: ^3.1.1
+  test: ^1.25.8
+  yaml: ^3.1.2
```
[the remark this answers is not in the document]
Comment (mosuem) [reply pkgs/native_assets_cli/pubspec.yaml:27]: Yeah, I also thought about this. I believe that if we want users to test their build hooks, then whoever depends on `package:hook` will also depend on `package:hook_test`, so it doesn't make a difference if we split it into a separate package.

File: pkgs/native_assets_cli/example/build/local_asset/test/build_test.dart [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    description: 'test my build hook',
+    mainMethod: build.main,
+    check: (output) {
+      expect(output.assets, isNotEmpty);
+      expect(output.assets.first, isA<NativeCodeAsset>());
+      expect(
+        (output.assets.first as NativeCodeAsset).id,
+        'package:local_asset/asset.txt',
+      );
+    },
```
[the remark this answers is not in the document]
Comment (mosuem) [reply pkgs/native_assets_cli/example/build/local_asset/test/build_test.dart:21]: Yes, in follow-up PRs ;)

We also need a `testLinkHook`.

**Review by dcharkes — APPROVED** [review]

**Pull request merged**